### PR TITLE
Restore backed up data in PVC Template

### PR DESCRIPTION
**Pull request by suaas21**
Task:
- [x] Advance Use Case
  - [x] Restore volumes in PVC template
  - [x]  Instant backup

xref: #https://github.com/stashed/stash/issues/706

### CODE REVIEW

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+```console
+$ kubectl create ns demo
+namespace/demo created
+```
+
+>**Note:** YAML files used in this tutorial are stored in  [docs/examples/guides/latest/advance-use-case/instant-backup](/docs/examples/guides/latest/advance-use-case/instant-backup) directory of [stashed/stash](https://github.com/stashed/stash) repository.
+
+## Backup Configure
+
+Here, we are going to backup the data volumes of Deployment. At first, we will deploy a Deployment with a PVC and generate some sample data in it. Then we will take instant backup of these PVC using Stash.
```
[the remark this answers is not in the document]
Comment (hossainemruz) [reply docs/guides/latest/advance-use-case/instant-backup.md]: Then, we will configure backup for the Deployment using Stash.

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+**Deploy Deployment:**
+
+We will deploy a Deployment with PVC. This Deployment will automatically generate sample data in `/source/data` directory where we have mounted the PVC.
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: remove this

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        volumeMounts:
+        - mountPath: /source/data
+          name: source-data
+      restartPolicy: Always
+      volumes:
+      - name: source-data
+        persistentVolumeClaim:
+          claimName: source-data
+```
+
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: The above Deployment will automatically create a `data.txt` file in `/source/data` directory and write some sample data in it.

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+Let’s create the Deployment and PVC we have shown above.
+
+```console
+$ kubectl apply -f ./docs/examples/guides/latest/advance-use-case/instant-backup/deployment.yaml
+persistentvolumeclaim/source-data created
+deployment.apps/stash-demo created
+```
+
+Now, wait for the pod of Deployment to go into the `Running` state.
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: Now, wait for pod of the Deployment to go into `Running` state.

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+Verify that the sample data has been created in `/source/data` directory using the following command,
+
+```console
+$ kubectl exec -n demo stash-demo-859d96f6bd-fxr7l ls /source/data
+data.txt
+```
+
+**Create Secret and Repository:**
+
+We are going to store our backed up data into a GCS bucket using Stash. We have to create a Secret and a Repository object with access credentials and backend information respectively.
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: Remove `using Stash`

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+
+```console
+$ kubectl get pod -n demo
+NAME                         READY   STATUS    RESTARTS   AGE
+stash-demo-9bff9fd4f-xvt77   2/2     Running   0          57s
+```
+
+Look at the pod. It now has 2 containers. If you view the resource definition of this pod, you will see that there is a container named `stash` which is running `run-backup` command.
+
+```yaml
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: remove this

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+
+Verify that the `CronJob` has been created using the following command,
+
+```console
+$ kubectl get backupconfiguration -n demo 
+NAME                TASK   SCHEDULE       PAUSED   AGE
+deployment-backup          */40 * * * *            6m41s
+```
+
+## Backup Instant
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: ## Trigger Instant Backup

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+
+Now, we will take instant backup of the data volumes. for that, we will deploy a `BackupSession` object to trigger instant backup.
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: Now, we will trigger an instant backup of Deployment's volumes that we have configured in the previous section. In order to do that, we have to create a `BackupSession` object pointing to the respective `BackupConfiguration` object.

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+  labels:
+    stash.appscode.com/backup-configuration: deployment-backup
+  name: deployment-backupsession
+  namespace: demo
+spec:
+  backupConfiguration:
+    name: deployment-backup
+```
+
+- `metadata.labels` holds the respective `BackupConfiguration` name as label. Stash backup sidecar container use this label to watch only the BackupSessions of that `BackupConfiguration`. The form of label must follow the following convention, `stash.appscode.com/backup-configuration: <BackupConfiguration name>`.
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: You must provide the label in the following format:
```
stash.appscode.com/backup-configuration: <BackupConfiguration name>
```

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: Explain `spec.backupconfiguration.name` field.

File: docs/guides/latest/advance-use-case/instant-backup.md [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+Let's create the `BackupSession` object that we have have shown above,
+
+```console
+$ kubectl apply -f ./docs/examples/guides/latest/advance-use-case/instant-backup/backupsession.yaml
+backupsession.stash.appscode.com/deployment-backupsession created
+```
+
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/instant-backup.md]: If everything goes well, the `stash` sidecar inside the Deployment will take a backup instantly.

File: docs/guides/latest/advance-use-case/restore-in-pvc-template.md [hunk 12]
```diff
@@ -0,0 +1,831 @@
...
+
+Now, wait for pod of the deployment to go into the `Running` state.
+
+```console
+$ kubectl get pod -n demo
+NAME                                READY   STATUS    RESTARTS   AGE
+restore-demo-85fbcb5dcf-vpbt8       1/1     Running   0          2m50s
+```
+
+Verify that the sample data has been created in `/source/data` and `/source/config` directory using the following command,
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/restore-in-pvc-template.md]: Verify that the sample data has been created -> Verify that the backed up data has been restored

File: docs/guides/latest/advance-use-case/restore-in-pvc-template.md [hunk 13]
```diff
@@ -0,0 +1,831 @@
...
+        accessModes: [ "ReadWriteOnce" ]
+        storageClassName: "standard"
+        resources:
+          requests:
+            storage: 2Gi
+```
+
+- `spec.target.replicas` `spec.target.replicas` specify the number of replicas of a StatefulSet whose volumes was backed up and Stash uses this field to dynamically create the desired number of PVCs and initialize them from respective Volumes.
+- `spec.target.volumeClaimTemplates:` specifies a list of PVC templates that will be created by restoring volumes from respective backed up volumes.
+  - `metadata.name` Specifies the name of the restored PVC. You must provide the name in the format: `pvcname-<POD_ORDINAL>`. `${POD_ORDINAL}` is resolved by the stash operator.
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/restore-in-pvc-template.md]: explain similarly to VolumeSnapshot doc.

File: docs/guides/latest/advance-use-case/restore-in-pvc-template.md [hunk 14]
```diff
@@ -0,0 +1,831 @@
...
+
+```console
+$ kubectl get pod -n demo 
+NAME             READY   STATUS    RESTARTS   AGE
+restore-demo-0   1/1     Running   0          34s
+restore-demo-1   1/1     Running   0          31s
+restore-demo-2   1/1     Running   0          26s
+```
+
+Verify that the sample data has been created in `/restore/data` and `/restore/config` directory using the following command,
```
Comment (hossainemruz) [inline docs/guides/latest/advance-use-case/restore-in-pvc-template.md]: Show restored data for each pod

Comment (hossainemruz) [reply docs/guides/latest/advance-use-case/restore-in-pvc-template.md]: Use `cat`

**Comment by tamalsaha** [comment]
I don't think these are advanced use-cases.

File: docs/guides/latest/advance-use-case/clone-pvc.md [hunk 15]
```diff
@@ -0,0 +1,851 @@
...
+apiVersion: stash.appscode.com/v1beta1
+kind: RestoreSession
+metadata:
+  name: restore-deployment
+  namespace: demo
+spec:
+  repository:
+    name: gcs-repo
+  rules:
+    - paths:
```
Comment (tamalsaha) [inline docs/guides/latest/advance-use-case/clone-pvc.md]: Should we call this directories?

**Pull request merged**